### DISCUSSION TITLE
Documenting attribute array arguments

### DISCUSSION THREAD

**Discussion by mlstowell** [mfem/mfem#2828] [General]
MFEM has two standard means of passing element/boundary attribute sets through function calls.
1. An array containing the relevant attribute numbers.
2. A marker array of length `Mesh::attributes.Max()` or `Mesh::bdr_attributes.Max()` where non-zero entries identify relevant attributes.

It is not always obvious which of these schemes is required by a particular function. I would like to add a concise statement to each relevant function which would make this clear.  Before making any changes to the documentation I'd like to discuss the best approach here and find a consensus.

**Comment by mlstowell** [comment]
To kick things off I'll make the following suggestion. We could add one of these short comments:
1. `set of attributes, see XXX for more information.`
2. `marker array, see XXX for more information.`

We would then need to add a more detailed explanation someplace. Perhaps in the `mesh.hpp` header where the attribute arrays are stored. Or with the documentation for a new function such as `Mesh::AttrToMarker` (a similar function is currently defined in `miniapps/common/mesh_extrs.hpp`).

**Comment by tzanio** [comment]
I very much agree that adding this documentation will be useful, thanks for suggesting in @mlstowell !

> 1. set of attributes, see XXX for more information.
> 2. marker array, see XXX for more information.

Both of these sound good to me. 

I like the idea of adding `Mesh::AttrToMarker` and putting the documentation there.

**Comment by v-dobrev** [comment]
Documenting attribute array usage systematically is definitely a good thing.

Note that we already have these static methods: https://github.com/mfem/mfem/blob/5bf8d8a5d137103b889f3f71efc11d4cb3ce2abb/fem/fespace.hpp#L796-L803 Actually, these can be used only for dofs since they use zero-based indexing (which is what we use for dofs) and we use 1-based indexing for the attributes. One possibility is to extend these methods with an optional parameter, `int base = 0`.

Regarding the terminology 1 and 2: some places in the code may use the term "list of attributes" meaning the same as "set of attributes". We can probably use these two interchangeably since the ordering implied by "list" is probably not important in most cases. The "marker array" is already used in some places, I think, at least in the context of dofs. For clarity, we may need to explicitly refer to 2 as "attribute marker array" to distiinguish it from "dofs marker array", to avoid potential confusion.

Another context where we use such arrays is for mesh elements (`Array<int>` or `Array<Refinement>`) for local refinement -- in these cases we use just "list" / "set" arrays, if I remember correctly.

**Reply by tzanio** [reply]
Good points.

I like `list of indices/attributes/dofs` and `attribute/dof/index marker array`.

Technically we encode these in an `Array`, which is ordered, so "list" may be better than "set".
